File: v1/net/encoding.py
"""
Neural input/output helpers for tensorized self-play.

This module will host `states_to_tensor` (batched) and utilities to project the
model logits back onto legal move masks according to the action encoding spec.
"""

from __future__ import annotations

from typing import Tuple

try:
    import v0_core  # type: ignore[attr-defined]
except ImportError:  # pragma: no cover - optional dependency
    v0_core = None

import torch

from ..game.move_encoder import ActionEncodingSpec, DEFAULT_ACTION_SPEC, DIRECTIONS
from ..game.state_batch import TensorStateBatch
from .project_policy_logits_fast import project_policy_logits_fast

_PHASE_ORDER = (
    1,  # PLACEMENT
    2,  # MARK_SELECTION
    3,  # REMOVAL
    4,  # MOVEMENT
    5,  # CAPTURE_SELECTION
    6,  # FORCED_REMOVAL
    7,  # COUNTER_REMOVAL
)
_NUM_PHASE_CHANNELS = len(_PHASE_ORDER)
_NUM_BOARD_CHANNELS = 2
_NUM_MARK_CHANNELS = 2
_TOTAL_CHANNELS = _NUM_BOARD_CHANNELS + _NUM_MARK_CHANNELS + _NUM_PHASE_CHANNELS


def states_to_model_input(batch: TensorStateBatch) -> torch.Tensor:
    """
    Convert a `TensorStateBatch` into model-ready tensors (B, C, H, W).
    """
    if not isinstance(batch, TensorStateBatch):
        raise TypeError(f"states_to_model_input expects TensorStateBatch, got {type(batch)!r}")

    if v0_core is not None and hasattr(v0_core, "states_to_model_input"):
        try:
            return v0_core.states_to_model_input(
                batch.board,
                batch.marks_black,
                batch.marks_white,
                batch.phase,
                batch.current_player,
            )
        except RuntimeError:
            pass

    board = batch.board
    if board.dim() != 3:
        raise ValueError(f"Expected board tensor of shape (B, H, W), got {tuple(board.shape)}")

    batch_size, height, width = board.shape
    device = board.device
    dtype = torch.float32

    board_planes = _board_planes(batch, dtype)
    mark_planes = _mark_planes(batch, dtype)
    phase_planes = _phase_planes(batch, dtype, height, width)

    stacked = torch.cat((board_planes, mark_planes, phase_planes), dim=1)
    if stacked.shape[1] != _TOTAL_CHANNELS:
        raise RuntimeError(
            f"states_to_model_input produced {stacked.shape[1]} channels; expected {_TOTAL_CHANNELS}"
        )
    return stacked.contiguous()


def _board_planes(batch: TensorStateBatch, dtype: torch.dtype) -> torch.Tensor:
    board = batch.board
    current = batch.current_player.view(-1, 1, 1).to(board.dtype)

    self_mask = (board == current).to(dtype)
    opp_mask = (board == -current).to(dtype)

    return torch.stack((self_mask, opp_mask), dim=1)


def _mark_planes(batch: TensorStateBatch, dtype: torch.dtype) -> torch.Tensor:
    current_is_black = (batch.current_player == 1).view(-1, 1, 1)

    self_marks = torch.where(current_is_black, batch.marks_black, batch.marks_white).to(dtype)
    opp_marks = torch.where(current_is_black, batch.marks_white, batch.marks_black).to(dtype)

    return torch.stack((self_marks, opp_marks), dim=1)


def _phase_planes(
    batch: TensorStateBatch,
    dtype: torch.dtype,
    height: int,
    width: int,
) -> torch.Tensor:
    device = batch.phase.device
    phase_ids = torch.tensor(_PHASE_ORDER, dtype=batch.phase.dtype, device=device)
    matches = batch.phase.view(-1, 1) == phase_ids.view(1, -1)
    phase_one_hot = matches.to(dtype).view(-1, _NUM_PHASE_CHANNELS, 1, 1)
    return phase_one_hot.expand(-1, -1, height, width)


def _project_policy_logits_python(
    logits: Tuple[torch.Tensor, torch.Tensor, torch.Tensor],
    legal_mask: torch.BoolTensor,
    spec: ActionEncodingSpec = DEFAULT_ACTION_SPEC,
) -> Tuple[torch.Tensor, torch.Tensor]:
    """
    Combine raw logits with legal-action masks.

    Returns tuple `(probs, masked_logits)` for downstream sampling/training.
    """
    if len(logits) != 3:
        raise ValueError("project_policy_logits expects a tuple of (log_p1, log_p2, log_pmc).")

    log_p1, log_p2, log_pmc = logits
    if log_p1.shape != log_p2.shape or log_p1.shape != log_pmc.shape:
        raise ValueError("All policy heads must share the same shape.")

    batch_size, head_dim = log_p1.shape
    device = log_p1.device
    dtype = log_p1.dtype

    placement_dim = spec.placement_dim
    movement_dim = spec.movement_dim
    selection_dim = spec.selection_dim
    auxiliary_dim = spec.auxiliary_dim

    board_size = int(round(placement_dim ** 0.5))
    if board_size * board_size != placement_dim:
        raise ValueError("placement_dim must be a perfect square representing the board area.")

    if head_dim != placement_dim:
        raise ValueError(
            f"Policy head dimension mismatch: expected {placement_dim}, got {head_dim}."
        )
    total_dim = spec.total_dim
    if legal_mask.shape != (batch_size, total_dim):
        raise ValueError(
            f"legal_mask expected shape ({batch_size}, {total_dim}), got {tuple(legal_mask.shape)}"
        )

    neginf = torch.tensor(float("-inf"), device=device, dtype=dtype)

    placement_slice = slice(0, placement_dim)
    movement_slice = slice(placement_slice.stop, placement_slice.stop + movement_dim)
    selection_slice = slice(movement_slice.stop, movement_slice.stop + selection_dim)
    auxiliary_slice = slice(selection_slice.stop, selection_slice.stop + auxiliary_dim)

    combined = torch.empty((batch_size, total_dim), dtype=dtype, device=device)
    combined[:, placement_slice] = log_p1
    combined[:, selection_slice] = log_pmc

    dirs = len(DIRECTIONS)
    if movement_dim != placement_dim * dirs:
        raise ValueError(
            f"movement_dim mismatch: expected {placement_dim * dirs}, got {movement_dim}."
        )

    indices = torch.arange(placement_dim, device=device)
    rows = indices // board_size
    cols = indices % board_size

    movement_logits = []
    for dr, dc in DIRECTIONS:
        dest_rows = rows + dr
        dest_cols = cols + dc
        valid = (
            (dest_rows >= 0)
            & (dest_rows < board_size)
            & (dest_cols >= 0)
            & (dest_cols < board_size)
        )
        dest_indices = dest_rows * board_size + dest_cols
        dest_indices = dest_indices.to(torch.long)

        dest_scores = torch.full((batch_size, placement_dim), neginf, dtype=dtype, device=device)
        if valid.any():
            gather_idx = dest_indices[valid].view(1, -1).expand(batch_size, -1)
            gathered = log_p1.gather(1, gather_idx)
            dest_scores[:, valid] = gathered

        movement_logits.append(log_p2 + dest_scores)

    movement_concat = torch.stack(movement_logits, dim=2).reshape(batch_size, movement_dim)
    combined[:, movement_slice] = movement_concat

    if auxiliary_dim > 0:
        aux = torch.zeros((batch_size, auxiliary_dim), dtype=dtype, device=device)
        combined[:, auxiliary_slice] = aux

    masked_logits = torch.where(legal_mask, combined, neginf)
    probs = torch.zeros_like(combined)

    for idx in range(batch_size):
        legal_indices = legal_mask[idx].nonzero(as_tuple=False).view(-1)
        if legal_indices.numel() == 0:
            continue
        row_logits = masked_logits[idx, legal_indices]
        if not torch.isfinite(row_logits).any():
            row_logits = torch.zeros_like(row_logits)
            masked_logits[idx, legal_indices] = row_logits
        if legal_indices.numel() == 1:
            probs[idx, legal_indices[0]] = 1.0
        else:
            row_probs = torch.softmax(row_logits, dim=0)
            probs[idx, legal_indices] = row_probs

    return probs, masked_logits


def project_policy_logits(
    logits: Tuple[torch.Tensor, torch.Tensor, torch.Tensor],
    legal_mask: torch.BoolTensor,
    spec: ActionEncodingSpec = DEFAULT_ACTION_SPEC,
) -> Tuple[torch.Tensor, torch.Tensor]:
    """
    Combine raw logits with legal-action masks, preferring the fast C++ path when available.

    Returns tuple `(probs, masked_logits)` for downstream sampling/training.
    """
    if len(logits) != 3:
        raise ValueError("project_policy_logits expects a tuple of (log_p1, log_p2, log_pmc).")

    log_p1, log_p2, log_pmc = logits
    try:
        fast_result = project_policy_logits_fast(log_p1, log_p2, log_pmc, legal_mask, spec)
    except RuntimeError:
        fast_result = None

    if fast_result is not None:
        return fast_result

    return _project_policy_logits_python(logits, legal_mask, spec)

File: v0/src/net/encoding.cpp
#include "v0/net_encoding.hpp"

#include <algorithm>
#include <array>
#include <limits>

namespace v0 {

namespace {

constexpr std::array<int64_t, 7> kPhaseOrder = {1, 2, 3, 4, 5, 6, 7};

torch::Tensor ensure_contiguous(const torch::Tensor& tensor) {
    return tensor.is_contiguous() ? tensor : tensor.contiguous();
}

int64_t resolve_dim(int64_t requested_dim, int64_t total_dims) {
    if (requested_dim >= 0) {
        return requested_dim;
    }
    return total_dims + requested_dim;
}

}  // namespace

torch::Tensor states_to_model_input(
    const torch::Tensor& board_in,
    const torch::Tensor& marks_black_in,
    const torch::Tensor& marks_white_in,
    const torch::Tensor& phase_in,
    const torch::Tensor& current_player_in) {
    TORCH_CHECK(board_in.dim() == 3, "board must be (B, H, W)");
    auto board = ensure_contiguous(board_in);
    auto marks_black = ensure_contiguous(marks_black_in);
    auto marks_white = ensure_contiguous(marks_white_in);
    auto phase = ensure_contiguous(phase_in);
    auto current_player = ensure_contiguous(current_player_in);

    const int64_t batch_size = board.size(0);
    const int64_t height = board.size(1);
    const int64_t width = board.size(2);
    TORCH_CHECK(marks_black.sizes() == board.sizes(), "marks_black shape mismatch");
    TORCH_CHECK(marks_white.sizes() == board.sizes(), "marks_white shape mismatch");
    TORCH_CHECK(phase.numel() == batch_size, "phase length mismatch");
    TORCH_CHECK(current_player.numel() == batch_size, "current_player length mismatch");

    auto dtype = torch::kFloat32;
    auto device = board.device();

    auto board_dtype = board.scalar_type();
    auto current = current_player.view({batch_size, 1, 1}).to(board_dtype);
    auto self_mask = board.eq(current).to(dtype);
    auto opp_mask = board.eq(current.neg()).to(dtype);
    auto board_planes = torch::stack({self_mask, opp_mask}, 1);

    auto marks_black_f = marks_black.to(dtype);
    auto marks_white_f = marks_white.to(dtype);
    auto current_is_black = current_player.eq(1).view({batch_size, 1, 1});
    auto self_marks = torch::where(current_is_black, marks_black_f, marks_white_f);
    auto opp_marks = torch::where(current_is_black, marks_white_f, marks_black_f);
    auto mark_planes = torch::stack({self_marks, opp_marks}, 1);

    // auto phase_ids = torch::tensor(
    //     kPhaseOrder,
    //     torch::TensorOptions().dtype(torch::kInt64).device(phase.device()));

    auto phase_ids = torch::arange(
        /*start=*/1,
        /*end=*/static_cast<int64_t>(kPhaseOrder.size()) + 1,
        torch::TensorOptions().dtype(torch::kInt64)
    ).to(phase.device());
    auto matches =
        phase.view({batch_size, 1}).eq(phase_ids.view({1, static_cast<int64_t>(kPhaseOrder.size())}));
    auto phase_one_hot = matches.to(dtype).view({batch_size, static_cast<int64_t>(kPhaseOrder.size()), 1, 1});
    auto phase_planes = phase_one_hot.expand({batch_size, static_cast<int64_t>(kPhaseOrder.size()), height, width});

    auto stacked = torch::cat({board_planes, mark_planes, phase_planes}, 1).contiguous();
    return stacked;
}

torch::Tensor postprocess_value_head(const torch::Tensor& raw_values) {
    return torch::tanh(raw_values);
}

torch::Tensor apply_temperature_scaling(
    const torch::Tensor& probs_in,
    double temperature,
    int64_t dim) {
    auto probs = probs_in.contiguous();
    if (temperature <= 1e-6) {
        return probs.clone();
    }
    auto exponent = 1.0 / std::max(temperature, 1e-6);
    auto positive = torch::where(
        probs > 0,
        probs.pow(exponent),
        torch::zeros_like(probs));

    int64_t reduce_dim = resolve_dim(dim, probs.dim());
    TORCH_CHECK(reduce_dim >= 0 && reduce_dim < probs.dim(), "Invalid dimension for temperature scaling");
    auto sums = positive.sum(reduce_dim, /*keepdim=*/true);
    auto normalized = torch::where(
        sums > 0,
        positive / sums,
        torch::zeros_like(positive));
    return normalized;
}

}  // namespace v0
